mono/metadata/icall.c

```diff
@@ -7391,6 +7391,28 @@ ves_icall_System_Environment_Exit (int result)
 	exit (result);
 }
 
+void
+ves_icall_System_Environment_FailFast (MonoStringHandle message, MonoExceptionHandle exception, MonoStringHandle errorSource, MonoError *error)
+{
+	if (MONO_HANDLE_IS_NULL (message)) {
+		g_warning ("CLR: Managed code called FailFast without specifying a reason.");
+	} else {
+		char *msg = mono_string_handle_to_utf8 (message, error);
+		g_warning ("CLR: Managed code called FailFast, saying \"%s\"", msg);
+		g_free (msg);
```

Comment:
On Windows please consider `%ls` and passing through the Unicode.
i.e. no alloc, copy, free.
For that matter, if `sizeof(wchar_t) == sizef(gunichar2)`, maybe something.

Reply:
Also please consider a small, maybe 64 local buffer, init with 0 and copy the first 64 or length chars from message.
i.e. if a dump is made, and includes stack, but doesn't follow pointers -- I don't know.

Reply:
Agreed on both points. I just wanted to have something working first, pass the CI and then improve upon that.

Reply:
Looks good despite all my nitpicking.

Reply:
Shouldn't this use `syslog` to emulate Event Log semantics on WIndows?

Reply:
This is not the part that is supposed to be written to event log. The event log should be written roughly at the `abort ();` call. In fact it happens on Windows that way and you get link to the WER minidump if enabled.

Printing this message is largely supposed to offer some indication either through attached native debugger, or through system log if no debugger is attached. On CoreCLR this goes through `OutputDebugString` on Windows. `syslog` may or may not be a good substitute, not sure about it.

`g_warning` is not perfect but it gets the job done here (both developer and user can see the message) and it has the added bonus of being wired properly on Android builds.